### ISSUE TITLE
Can a constructor named `runtimeType` or `hashCode` be torn off?

### ISSUE THREAD

**Issue by srawlins** [dart-lang/language#1718]
Forgive my bare understanding of the parsing rules, but I was surprised to see co19 language tests like [named_constructor_A03_t02](https://github.com/dart-lang/co19/blob/master/LanguageFeatures/Constructor-tear-offs/named_constructor_A03_t02.dart) which include code like:

```dart
var v3 = (C<dynamic>).constr;
```

I had thought that `C<dynamic>` when not followed immediately by a `.` would be a Type, which does not have a getter called `constr`. In any case, assuming this is a valid test, I wonder then, given a class with constructors `runtimeType` or `hashCode`, are the following variables constructor tearoffs or field accesses on `Type`?

```dart
class C<T> {
  C.runtimeType();
  C.hashCode();
}

void main() {
  (C).runtimeType; // In current Dart, this accesses `runtimeType` on the Type.
  (C<int>).runtimeType; // Illegal in current Dart.
}
```

In other words, do constructors shadow instance getters that exist on `Type`, or vice versa?

I think the question can extend to a constructor named `toString`. `(C).toString` normally represents the tearoff of the `toString` function on `Type`, but what if `C` has a constructor named `toString`?

**Comment by lrhn** [comment]
For non-generic types, the answer is simply "yes", they can be torn off.

Also, I think `(C<int>).constr` should be a compile-time error.

The current behavior for `C.runtimeType` is to say that `C` doesn't have a static member named `runtimeType`. You already need to write `(C).runtimeType` to get the `Type` object for `Type` (Or just write `Type`, since you know what the result will be).
With constructor tear-offs, that expression, which is currently invalid, would become a valid constructor tear-off.

For `C<int>.runtimeType`, which is currently invalid, we need to consider our options. 
You cannot access static members as `C<int>.staticOnC` (we could allow it, but it looks deceptively like the static members could use the type variable). That opens us up to allowing `C<int>.instanceOnType`, but we don't actually *need* that (there aren't that many interesting instance members on `Type`) and it's inconsistent with the non-generic case, so I'd go with disallowing it, and only allowing `C<int>.something` when `something` is a constructor. You'll still need `(C<int>).toString()` to get `"C"` (again, you know the result, so it's a useless expression to give preference to).

And `(typeLiteral).instanceMemberOnType` should still give you access to instance members on `Type`, nothing more. Static/constructor accesses need to be directly on the class, not parenthesized.

**Comment by srawlins** [comment]
Thanks!

> Also, I think `(C<int>).constr` should be a compile-time error.

Then tests like `named_constructor_A03_t02` which use it need to be fixed? Or is that notion just held in the test status database?

> And `(typeLiteral).instanceMemberOnType` should still give you access to instance members on `Type`, nothing more. Static/constructor accesses need to be directly on the class, not parenthesized.

Perfect. The ambiguity I was worried about vanishes if `(C<int>).constr` is a compile-time error.

**Comment by eernstg** [comment]
> tests like `named_constructor_A03_t02` which use it need to be fixed?

Yes, that one should expect the compile-time error that `Type` has no member named `constr` at lines 40, 44, 48, or it should stop using those `()` such that the test can be executed.

**Comment by srawlins** [comment]
@lrhn wrote:

> Also, I think `(C<int>).constr` should be a compile-time error.

I think the spec contains two examples of that syntax:

> `var v5 = (C.name)<typeArgs>(args);`

> You can write `(List.filled)<int>(4, 4)` to do the tear-off

**Comment by lrhn** [comment]
That's slightly different syntax.

You can do the tear-off (which is what the parentheses ensure) then invoke the resulting generic function.
You *cannot* instantiate a function value, nor can you tear-off constructors from a `Type` object, which is why
`(C.named)<int>` and `(C<int>).named` are both invalid (the former only if not followed by `(...)`, because then it's a call, not an instantiation).

Assume we have a generic `class C<T>` with an unnamed constructor `C();` and one named `C.name();`, then

| expression | valid? | meaning |
|:-------------|:-----:|:------------|
| C                | &checkmark; | Type literal for `C<dynamic>` |
| C\<int>       | &checkmark; | Type literal for `C<int>` |
| (C)\<int>     |&cross; | Cannot instantiate a type literal |
| C.named <br> C.new |&checkmark; | Generic constructor tear-off |
| (C).named | &cross; | No `named` member on `Type` |
| (C).new | &cross; | Not even valid syntax, only `new` after a *class* |
| C\<int>.named <br> C\<int>.new| &checkmark; | Instantiated constructor tear-off |
| (C\<int>).named | &cross; | No `named` member on `Type` |
| (C\<int>).new | &cross; | Still not even valid syntax |
| C.named\<int> <br> C.new\<int> | &cross; | The constructor is not generic. |
| (C.named)\<int> <br> (C.new)\<int> | &cross; | Cannot instantiate a function value |
| C()  | &checkmark; | Unnamed constructor invocation, implicit type argument |
| (C)() | &cross; | `Type` object isn't callable |
| C.named() <br> C.new() | &checkmark; | Generic constructor invocation with implicit type arguments |
| (C.named)() <br> (C.new)() | &checkmark; | Generic constructor tear-off, invoked with implicit type arguments |
| (C).named() <br> (C).new() | &cross; | Not valid on a `Type` object |
| C\<int>() | &checkmark; | Unnamed constructor invocation, explicit type argument |
| (C\<int>)() | &cross; | Cannot call a `Type` object |
| (C)\<int>() | &cross; | Cannot call a `Type` object |
| C\<int>.named() <br> C\<int>.new() | &checkmark; | Constructor invocation, explicit type argument |
| (C\<int>.named)() <br> (C\<int>.new)() | &checkmark; | Instantiated constructor tear-off, invoked |
| (C\<int>).named() <br> (C\<int>).new() | &cross; | Not valid on a `Type` object |
| (C)\<int>.named() <br> (C)\<int>.new() | &cross; | Not valid on a `Type` object |

In general, you can only put `<int>` after a reference to a declaration (or instance method) unless it's followed by `(...)`, then it's a generic invocation. Anything parenthesized is a *value*, not a declaration.

**Comment by srawlins** [comment]
Oops, my mistake is  obvious, haha. Thanks for the great explanation, @lrhn !
